### ISSUE TITLE
Contrast Checker only works after changing colors the second time

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#60825]
### Description

Discovered while exploring https://github.com/WordPress/gutenberg/issues/58936#issuecomment-2060513454

It appears the Contrast Checker is buggy and only works when changing _one more time_ a color setting. This happens both when setting an insufficient color pair *the notice should appear) and when setting a sufficient one (the notice should disappear).

See attached animated GIF.

The Contrast Checker is an important feature that should work 'out of the box' for all color settings that apply to text and background. Since WordPress 6.6 is supposed to be a 'bug fixes' release, I'd like to propose to fix the Contrast Checker and make it work reliably for 6.6. Some e2e tests would be nice.

![initial](https://github.com/WordPress/gutenberg/assets/1682452/83d14688-37a6-436c-8917-1401c037f985)

### Step-by-step reproduction instructions

- Create a new post.
- Add a Paragraph block with some text.
- In the Settings panel, set both the text and background colors to the same color.
- Observe the Contrast Checker doesn't render any notice.
- Change one of the colors to another one that still produces an insufficient color pair.
- Observe this time the Contrast Checker does render the notice.
- Set the color text to one that produces a _sufficient_ color pair contrast.
- Observee the Contrast Checker notice does not disappear and still warns about insufficient contrast. 
- Change the background color to one that still produces a _sufficient_ color pair contrast.
- Observe this time the Contrast Checker notice does disappear, as expected.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by afercia** [comment]
After some initial debugging, it appears the ContrastChcker does not (always) receive the changed colors. See attached GIF, while dumping to the console the colors [at the top of the file](https://github.com/WordPress/gutenberg/blob/3d8b192c3df3b02f4c1efc9d8154b715dc189600/packages/block-editor/src/components/contrast-checker/index.js#L42) in the ContrastChcker. 

![contrast](https://github.com/WordPress/gutenberg/assets/1682452/a55db9b1-ed7d-4dbf-b689-8f97169f451a)

**Comment by afercia** [comment]
Note to myself and to everyone willing to help on this issue.

The `ContrastChecker` component is used in a few different places (blocks, global styles edit palette, etc.). 

Some blocks, e.g. Social links, use `ContrastChecker` in a 'custom' way. For most of the blocks, the colors panel comes from [the color hook](https://github.com/WordPress/gutenberg/blob/4002fbdc8dc1e3170694ed47c6e7d30de760f98d/packages/block-editor/src/hooks/color.js#L326), which in turn uses the [BlockColorContrastChecker hook](https://github.com/WordPress/gutenberg/blob/4002fbdc8dc1e3170694ed47c6e7d30de760f98d/packages/block-editor/src/hooks/contrast-checker.js#L16), which is responsible to detect the colors.

As far as I can tell, it appears it's a matter of timing, as the color detection _sometimes_ happens before the actual color change. That is: `BlockColorContrastChecker` may return the _previous_ color.

See attached screenshot, where I'm dumping the detected text color, and dumping it again within a `0` setTimeout, for debugging purposes. See attached sample debugging code.

- `rgb(17, 17, 17)` is the initial text color wrongly returned by the hook.
- `rgb(216, 97, 60)` is the expected orangish new text color, which gets correctly detected only after the setTimeout.

![Screenshot 2024-05-07 at 15 36 39](https://github.com/WordPress/gutenberg/assets/1682452/54456441-4410-452e-ab0f-2271a383fa63)

[debug-contrast-checker.diff.txt](https://github.com/WordPress/gutenberg/files/15236674/debug-contrast-checker.diff.txt)

To debug:
- Apply the diff.
- Select a paragraph block and change the text color.
- Observe `BlockColorContrastChecker` detects the initial color. Expected: to detext the new color.
- Observe detecting the color within the setTimout returns the new color.

**Comment by afercia** [comment]
I can't figure out how to make sure the color detection is guaranteed to happen _after_ a color change. I spent a considerable amount of time debugging this issue but I'm afraid it goes beyond my coding abilities.

I would appreciate someone else more familiar with hooks to step in and help this issue move on. 
Cc @WordPress/gutenberg-core

**Comment by Mamaduka** [comment]
Closing in favor of #67035. I can no longer reproduce the bug due to different regressions. Details: https://github.com/WordPress/gutenberg/issues/67035#issuecomment-2479487449.
